### PR TITLE
[Expanded Weights] fix conv3d

### LINKED ISSUE #80953: ExpandedWeights fail on Conv3D
### 🐛 Describe the bug

`call_for_per_sample_grads` produces shape error if called on `nn.Conv3d` layer (at least in the configuration as below)

```python
N, C, D, H, W = 1, 32, 3, 6, 6
out_channels = C*32
kernel_size=2
stride=1
padding=2
dilation=1
groups=2

conv = nn.Conv3d(
    in_channels=C,
    out_channels=out_channels,
    kernel_size=kernel_size,
    stride=stride,
    padding=padding,
    dilation=dilation,
    groups=groups,
)
x = torch.randn([N, C, D, H, W])

# normal forward/backward works fine
out = conv(x)
out.sum().backward()

#ExpandedWeights crashes with shape error
out = call_for_per_sample_grads(conv, x.shape[0], x)
out.sum().backward()
```

Error:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
/var/folders/_r/hzffvgfd23sc3w0gr577_qmc0000gn/T/ipykernel_71302/1832167267.py in <module>
      1 #ExpandedWeights crashes with shape error
      2 out = call_for_per_sample_grads(conv, x.shape[0], x)
----> 3 out.sum().backward()

~/Documents/opacus/venv/lib/python3.7/site-packages/torch/_tensor.py in backward(self, gradient, retain_graph, create_graph, inputs)
    394                 create_graph=create_graph,
    395                 inputs=inputs)
--> 396         torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
    397 
    398     def register_hook(self, hook):

~/Documents/opacus/venv/lib/python3.7/site-packages/torch/autograd/__init__.py in backward(tensors, grad_tensors, retain_graph, create_graph, grad_variables, inputs)
    173     Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
    174         tensors, grad_tensors_, retain_graph, create_graph, inputs,
--> 175         allow_unreachable=True, accumulate_grad=True)  # Calls into the C++ engine to run the backward pass
    176 
    177 def grad(

~/Documents/opacus/venv/lib/python3.7/site-packages/torch/autograd/function.py in apply(self, *args)
    251                                "of them.")
    252         user_fn = vjp_fn if vjp_fn is not Function.vjp else backward_fn
--> 253         return user_fn(self, *args)
    254 
    255     def apply_jvp(self, *args):

~/Documents/opacus/venv/lib/python3.7/site-packages/torch/nn/utils/_expanded_weights/conv_expanded_weights.py in backward(ctx, grad_output)
     35     @staticmethod
     36     def backward(ctx, grad_output):
---> 37         return conv_backward(ctx.conv_fn, ctx, grad_output)

~/Documents/opacus/venv/lib/python3.7/site-packages/torch/nn/utils/_expanded_weights/conv_utils.py in conv_backward(func, ctx, grad_output)
     71 
     72     # set grad_sample field for weight and bias with per sample gradients
---> 73     set_grad_sample_if_exists(ctx.weight, weight_grad_sample)
     74     set_grad_sample_if_exists(ctx.bias, lambda _: grad_output.reshape(*grad_output.shape[:2], -1).sum(dim=2))
     75     return tuple(results)

~/Documents/opacus/venv/lib/python3.7/site-packages/torch/nn/utils/_expanded_weights/expanded_weights_utils.py in set_grad_sample_if_exists(maybe_expanded_weight, per_sample_grad_fn)
     64             unpacked.grad_sample = unpacked.grad_sample + per_sample_grad_fn(unpacked)
     65         else:
---> 66             unpacked.grad_sample = per_sample_grad_fn(unpacked)
     67 
     68 def unpack_expanded_weight_or_tensor(maybe_expanded_weight, func=lambda x: x):

~/Documents/opacus/venv/lib/python3.7/site-packages/torch/nn/utils/_expanded_weights/conv_utils.py in weight_grad_sample(weight)
     36         else:
     37             return conv_unfold_weight_grad_sample(ctx.input, grad_output, weight_shape, kernel_size,
---> 38                                                   stride, padding, dilation, groups, func)
     39 
     40     def expand(param):

~/Documents/opacus/venv/lib/python3.7/site-packages/torch/nn/utils/_expanded_weights/conv_utils.py in conv_unfold_weight_grad_sample(input, grad_output, weight_shape, kernel_size, stride, padding, dilation, groups, func)
    106     weight_grad_sample = torch.einsum("ngrg...->ngr...", weight_grad_sample).contiguous()
    107     shape = [n] + list(weight_shape)
--> 108     weight_grad_sample = weight_grad_sample.view(shape)
    109     return weight_grad_sample
    110 

RuntimeError: shape '[1, 1024, 16, 2, 2, 2]' is invalid for input of size 1990656
```

### Versions

PyTorch version: 1.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.4 (x86_64)
GCC version: Could not collect
Clang version: 12.0.5 (clang-1205.0.22.9)
CMake version: Could not collect
Libc version: N/A

Python version: 3.7.5 (default, Oct 22 2019, 10:35:10)  [Clang 10.0.1 (clang-1001.0.46.4)] (64-bit runtime)
Python platform: Darwin-21.5.0-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy==0.910
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.4
[pip3] pytorch-lightning==1.5.3
[pip3] torch==1.12.0
[pip3] torchaudio==0.12.0
[pip3] torchmetrics==0.6.0
[pip3] torchvision==0.13.0
[conda] Could not collect

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80944
* __->__ #80943
* #80895
* #80892

Conv3d has an ordering typo in its unfold (it needs a custom unfold since unfold3d doesn't exist in torch). This was caught by Opacus but not by us because dilation, padding, and stride always matched in the test cases. Conv3d has very few test cases since it doesn't have an OpInfo and we were skipping some by the nn module testing filtering. So this updated this filtering to add more of the common nn tests (one of which did fail without the change) 

Closes #80953

### CODE REVIEW

File: test/test_expanded_weights.py [hunk 1]
```diff
@@ -480,10 +480,17 @@ def test_context_manager_multiple_inputs(self, test_case, device):
...
+    if 'module_name' in t and t['module_name'] in supported_modules:
+        return True
+    if 'fullname' in t and any([module + "_" in t['fullname'] for module in supported_modules]):
+        return not('Conv' in t['fullname'] and 'pad' in t['fullname'])
+
 # TODO: Once all of these use ModuleInfo, replace with ModuleInfo tests
 # These currently use the legacy nn tests
 supported_modules = ['Linear', 'Conv1d', 'Conv2d', 'Conv3d', 'Embedding', 'LayerNorm', 'GroupNorm', 'InstanceNorm']
-supported_tests = [t for t in module_tests + new_module_tests if 'module_name' in t and t['module_name'] in supported_modules]
+supported_tests = [t for t in module_tests + new_module_tests if filter_supported_tests(t)]
```
Comment (zou3519) [inline test/test_expanded_weights.py:493]: (no action required) As this PR (and the last one) exposed, we're not really sure of how much coverage we actually have on expanded weights due to the legacy nn tests and OpInfo not having coverage. We probably want to identify the set of modules ExpandedWeights works over (you probably have this already) and then figure out if we do have the right amount of test coverage

**Review by zou3519 — APPROVED** [review]

File: test/test_expanded_weights.py [hunk 2]
```diff
@@ -480,10 +480,17 @@ def test_context_manager_multiple_inputs(self, test_case, device):
             raise unittest.SkipTest("Can't get per sample gradients for input of rank 1")
         test_case._do_test_multi_input(module, input)
 
+def filter_supported_tests(t):
+    supported_modules = ['Linear', 'Conv1d', 'Conv2d', 'Conv3d', 'Embedding', 'LayerNorm', 'GroupNorm', 'InstanceNorm']
+    if 'module_name' in t and t['module_name'] in supported_modules:
+        return True
+    if 'fullname' in t and any([module + "_" in t['fullname'] for module in supported_modules]):
+        return not('Conv' in t['fullname'] and 'pad' in t['fullname'])
+
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply test/test_expanded_weights.py:489]: Yes definitely! I'll take this on

File: test/test_expanded_weights.py [hunk 3]
```diff
@@ -480,10 +480,17 @@ def test_context_manager_multiple_inputs(self, test_case, device):
...
 # TODO: Once all of these use ModuleInfo, replace with ModuleInfo tests
 # These currently use the legacy nn tests
 supported_modules = ['Linear', 'Conv1d', 'Conv2d', 'Conv3d', 'Embedding', 'LayerNorm', 'GroupNorm', 'InstanceNorm']
```
Comment (jbschlosser) [inline test/test_expanded_weights.py:492]: Fair to say that this list should be prioritized for ModuleInfo entries from your perspective?

Comment (samdow) [reply test/test_expanded_weights.py:492]: Yes! It would be really helpful to avoid these sorts of user/filtering errors to have them in ModuleInfos

**Pull request closed without merging**